### DISCUSSION TITLE
OOP Improvement - new Dialog scope

### DISCUSSION THREAD

**Discussion by ericwood8** [dotnet/csharplang#7411] [Language Ideas]
I would like to see a new scope (similar to private, protected and public) called dialog on the class or a property or method that allows only a visibility to a single target object (whether view or class).  Other objects such as other classes could not see the scoped class/property/method.  I suggest the syntax be a tilde (~) character followed by the target object you want to share with. Since C# is case-sensitive then it would need to be a valid target name of matching spelling and case in the project.  

For all these examples, assume there is a valid class YYYY:

**Class Example:**

`dialog~YYYY class XXXXX 
{
   ...
}``

**Property Example:**

`public class AAAA
{
    dialog~YYYY void NNNN { get; set; }
}`

**Method Example:**

`public class AAAA
{
    dialog~YYYY void Accumulate(int count)
    {
        ...
    }
}`

**PRACTICAL USE CASES**
   I can see this especially helping XAML code between a view and a dedicated view model (for MVVM) so protected from incorrect usage.

For all these examples, assume there is a valid DisqualifiedBadge.xaml that contains: 
<UserControl x:Class="TruPredict.Controls.DisqualifiedBadge" ...>
And assume DisqualifiedBadge.xaml.cs code behind that is partial class and DisqualifiedBadgeViewModel.cs that is for MVVM.
   
**Dedicated View Model Example:**
`dialog~DisqualifiedBadge class DisqualifiedBadgeViewModel
{
        ...
}`

**Shared View Model Example (where the set deals with PropertyChanged events) :**
`public class BadgeViewModel
{
       private DynamicDataGridColumns _disqualifiedColumnsCollection;
        /// <summary> Columns are used in the main grid </summary>
        dialog~DisqualifiedBadge DynamicDataGridColumns DisqualifiedColumnsCollection
        {
            get { return _disqualifiedColumnsCollection; }
            set { Set(nameof(DisqualifiedColumnsCollection), ref _disqualifiedColumnsCollection, value); }
        }

        ...
}`

Overall - I am fine with a new name, instead of "Dialog".  I could not think of anything else.

**Comment by HaloFour** [comment]
See: #2073

You can enforce that kind of relationship with analyzers.

**Reply by KennethHoff** [reply]
Analyzers and Source Generators are - for good reason - recommended as the solution to _a lot_ of scenarios, but it's incredibly daunting to implement.

I spent ~2 hours last night implementing my very first analyzer - a  "You have to mark all properties as required" analyzer(including a CodeFix). It was actually relatively straight forward.

**Reply by CyrusNajmabadi** [reply]
> but it's incredibly daunting to implement.

Initially yes.  You have to learn the domain and the API.  But once learned, those skills can be continually reused.

It's far more effective to invest there than waiting on lang changes :)

**Reply by ericwood8** [reply]
>You can enforce that kind of relationship with analyzers.

Are you telling me build this OR are you telling me it is workable for ya'll to built it?

If you are wanting me to build this, then think about it this way:
I am a blacksmith asking a functioning tool shop that makes tools everyday to make me a hammer and the tool shop says "oh yes, well take some wood and fashion it into a wooden shaft and metal and fire up the forge to 3200 degrees and then pound the metal on the anvil for 2 hours and make your own hammer".

>It's far more effective to invest there than waiting on lang changes 

Maybe language changes should not be so hard.  Maybe the tool shop could build hammers and then add the hammer to its offerings for sale.

**Reply by KennethHoff** [reply]
What he's saying is this:

You can either make your own hammer using the materials you have at your shop - could probably be completed within the day - _or_ you could wait 3+ years because the tool shop has an insane backlog

**Reply by HaloFour** [reply]
> or you could wait 3+ years because the tool shop has an insane backlog

Or never, as most feature requests to the language are not accepted, especially those that only aim to add guardrails around the language as it is.  It'd exceptionally unlikely that a narrow case like this will bubble up to the level of being adopted, and analyzers were intentionally added to the Roslyn compiler to support developers in carving out their own rules for scenarios like this.

**Reply by KennethHoff** [reply]
>Or never, as most feature requests to the language are not accepted

I was going to mention that as well, but I chose not to for some reason(to make the analogy better?)

**Reply by CyrusNajmabadi** [reply]
> I am a blacksmith asking a functioning tool shop

Note: I really like this analogy as both blacksmiths and carpenters *routinely* will create their own one-off tools/jigs/etc. to solve problems for their domains :) 

As the Language/Compiler platform we very much like giving the facility and ability for devs to be able to scratch these domain itches themselves.  It's 100% fine to come and discuss here and ask for these things to be considered.  And, if we discover it would be hugely valuable, it is something we might do.  But, in the more likely event that it's too niche to go into the lang itself, analyzers solve the problem (albeit by spreading the costs over the community).

**Reply by ericwood8** [reply]
> one-off tools/jigs/etc. 

I see this as a universal tool.  Do you really see this as a one-off?

**Reply by CyrusNajmabadi** [reply]
> I see this as a universal tool. Do you really see this as a one-off?

Currently yes.  :)

**Reply by ericwood8** [reply]
>You can enforce that kind of relationship with analyzers.

I have never done an analyzer.  Sounds good for a temporary work around, but how does it deal with the fact that the proposed code cannot compile?  "dialog~YYYY" would not be valid syntax for C# code at the moment.

**Reply by CyrusNajmabadi** [reply]
The idea is that instead of creating new syntax, you just encode the info you want in an attribute.  like `[Dialog(typeof(YYYY))]`.  Then your analyzer sees those and enforces whatever rules you have for your domain.

**Reply by ericwood8** [reply]
Thanks Cyrus for the idea of using an attribute!
